### PR TITLE
Fixes in sphinx docs links

### LINKED ISSUE #8107: Testing URL links in docs & readme
## 🚀 Feature

it may easily happen that some external links are not valid anymore so we shall test them all...

### Motivation

detect broken links in the process

### Alternatives

load the URLs via bs4? (a web scraping tool)
assert that the text you get back contains an expected string?

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?
Fixes links referred to in #16234. 
Fixes #8107

#### Fixed the following links
- [x] All the 4 example links in callback [examples](https://pytorch-lightning.readthedocs.io/en/stable/extensions/callbacks.html#examples)
- [x] Two links the [lightning transformer page](https://pytorch-lightning.readthedocs.io/en/latest/ecosystem/transformers.html), in the seamless memory and optimization point. Both Deepspeed and FairScale shared links. 
- [x] QConfig link in the docstring for Quantization Aware Training [example](https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.callbacks.QuantizationAwareTraining.html#quantizationawaretraining)
- [x] Online-training for recurrent neural networks paper link in `common/lightning_module.rst` [Link](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_module.html#truncated-bptt-steps)
- [x] Efficient Data Batching link in docstrings was incorrect for [IPU Strategy](https://github.com/Lightning-AI/lightning/blob/ce5ea92b5372dd9ad157a16654f25c4ab2e2066e/src/pytorch_lightning/strategies/ipu.py#L67)
- [x] Opening reports with graph analyser in IPU Strategy [Link](https://pytorch-lightning.readthedocs.io/en/stable/accelerators/ipu_intermediate.html#popvision-graph-analyser) (last link on this page)
- [x] deepspeed3 offload link is incorrect in two places, [model parallel](https://pytorch-lightning.readthedocs.io/en/stable/advanced/model_parallel.html#deepspeed-zero-stage-3) and [fabric (prev lite)](https://pytorch-lightning.readthedocs.io/en/stable/starter/lightning_lite.html?highlight=deepspeed)
- [x] Updated the reference link for  `make_pruning_permanent()` in [ModelPruning Callback](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelPruning.html#pytorch_lightning.callbacks.ModelPruning.make_pruning_permanent)
- [x] Linkchecker fails on a few relative links (as listed below), added them to `lincheck_ignore` in `conf.py`
```
(common/hyperparameters: line   14) broken    ../cli/lightning_cli.html - 
(starter/introduction: line  182) broken    ../common/trainer.html#trainer-flags - 
(starter/introduction: line  102) broken    installation.html - 
(           index: line   65) broken    starter/installation.html - 
```

#### Fix the following links
```
(notebooks/lightning_examples/augmentation_kornia: line  128) broken    https://www.kornia.org - HTTPSConnectionPool(host='www.kornia.org', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:1131)')))
```

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@Borda @awaelchli This is ready for review. 

The [kornia.org](kornia.org) link in `notebooks/lightning_examples/augmentation_kornia` does not work (it is [kornia.github.io](kornia.github.io) now). When I run the linkcheck in my local, it shows this error, but it is showing that the build is successful [here](https://github.com/Lightning-AI/lightning/actions/runs/3850401392/jobs/6560707369#step:9:3738).

File: src/pytorch_lightning/callbacks/pruning.py [hunk 1]
```diff
@@ -261,7 +261,7 @@ def _wrap_pruning_fn(pruning_fn: Callable, **kwargs: Any) -> Callable:
     def make_pruning_permanent(self, module: nn.Module) -> None:
         """Removes pruning buffers from any pruned modules.
 
-        Adapted from https://github.com/pytorch/pytorch/blob/1.7.1/torch/nn/utils/prune.py#L1176-L1180
+        Adapted from https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/prune.py#L1172-L1200
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/pruning.py:264]: The specific version here is desired, as it was the exact code that was adapted back then. Otherwise, if pointing to master, it gets out of date.

File: docs/source-pytorch/starter/introduction.rst [hunk 2]
```diff
@@ -99,7 +99,7 @@ For `conda <https://anaconda.org/conda-forge/pytorch-lightning>`_ users
       </div>
    </div>
 
-Or read the `advanced install guide <installation.html>`_
+Or read the `advanced install guide <https://pytorch-lightning.readthedocs.io/en/stable/starter/installation.html>`_
```
Comment (awaelchli) [inline docs/source-pytorch/starter/introduction.rst:102]: Hard linking using the full url should be avoided whenever possible. We want that the links get generated. This applies to other places in this PR as well. Exceptions are in files that don't get generated by the docs (e.g., Readme files), or external urls to other websites.

In this specific example here, let's refer to installation.rst so the url gets properly generated.

File: docs/source-pytorch/index.rst [hunk 3]
```diff
@@ -62,7 +62,7 @@ Conda users
       </div>
    </div>
 
-Or read the `advanced install guide <starter/installation.html>`_
+Or read the `advanced install guide <https://pytorch-lightning.readthedocs.io/en/stable/starter/installation.html>`_
```
Comment (awaelchli) [inline docs/source-pytorch/index.rst:65]:
```suggestion
Or read the `advanced install guide <installation.rst>`_
```

File: src/pytorch_lightning/callbacks/pruning.py (hunk 1 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/callbacks/pruning.py:264]: Right, but what would be the solution if they have deleted the 1.7.1 branch from their repository?

File: docs/source-pytorch/starter/introduction.rst (hunk 2 above)
Comment (shenoynikhil) [reply docs/source-pytorch/starter/introduction.rst:102]: This raises a false positive error, I can do what you have suggested and include it in `linkcheck_ignore`.

File: docs/source-pytorch/index.rst (hunk 3 above)
Comment (shenoynikhil) [reply docs/source-pytorch/index.rst:65]: This raises a false positive error, I can do what you have suggested and include it in `linkcheck_ignore`.

File: src/pytorch_lightning/callbacks/pruning.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/pruning.py:264]: The branch is gone but the tag is still there. The difference is subtle: v1.7.1 vs 1.7.1 (the "v"):

https://github.com/pytorch/pytorch/blob/v1.7.1/torch/nn/utils/prune.py#L1176-L1180

Perhaps the code lines need to be adjusted

Comment (shenoynikhil) [reply src/pytorch_lightning/callbacks/pruning.py:264]: That makes sense. Is ths better
https://github.com/pytorch/pytorch/blob/v1.7.1/torch/nn/utils/prune.py#L1118-L1122

Comment (shenoynikhil) [reply src/pytorch_lightning/callbacks/pruning.py]: Resolved this.

File: docs/source-pytorch/accelerators/ipu_intermediate.rst [hunk 4]
```diff
@@ -60,4 +60,4 @@ Lightning supports dumping all reports to a directory to open using the tool.
     trainer = pl.Trainer(accelerator="ipu", devices=8, strategy=IPUStrategy(autoreport_dir="report_dir/"))
     trainer.fit(model)
 
-This will dump all reports to ``report_dir/`` which can then be opened using the Graph Analyser Tool, see `Opening Reports <https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html#opening-reports>`__.
+This will dump all reports to ``report_dir/`` which can then be opened using the Graph Analyser Tool, see `Opening Reports <https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/opening-reports.html>`__.
```
Comment (Borda) [inline docs/source-pytorch/accelerators/ipu_intermediate.rst:63]: does this need to be dropped while we shall be ignoring anchors?

File: docs/source-pytorch/starter/introduction.rst (hunk 2 above)
Comment (awaelchli) [reply docs/source-pytorch/starter/introduction.rst:102]: Yes, let's just ignore these false positives (that's one example why we didn't want a CI job that fails :))

**Review by awaelchli — APPROVED** [review]
LGTM after comments addressed

File: docs/source-pytorch/index.rst (hunk 3 above)
Comment (shenoynikhil) [reply docs/source-pytorch/index.rst]: Resolved. Reset them and added them to `linkcheck_ingore`.

File: docs/source-pytorch/starter/introduction.rst (hunk 2 above)
Comment (shenoynikhil) [reply docs/source-pytorch/starter/introduction.rst]: Resolved. Reset them and added them to `linkcheck_ingore`.

File: docs/source-pytorch/accelerators/ipu_intermediate.rst (hunk 4 above)
Comment (shenoynikhil) [reply docs/source-pytorch/accelerators/ipu_intermediate.rst:63]: Umm, I did not get the question. Even https://docs.graphcore.ai/projects/graph-analyser-userguide/en/latest/graph-analyser.html does not exist (ignoring the anchor).

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source-pytorch/conf.py [hunk 5]
```diff
@@ -412,3 +412,11 @@ def package_list_from_file(file):
 
 # ignore all links in any CHANGELOG file
 linkcheck_exclude_documents = [r"^(.*\/)*CHANGELOG.*$"]
+
+# ignore the following relative links
```
Comment (carmocca) [inline docs/source-pytorch/conf.py:416]: this comment should describe not the what, but the why. Why do we need to ignore these?

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source-pytorch/conf.py (hunk 5 above)
Comment (shenoynikhil) [reply docs/source-pytorch/conf.py]: Updated this with the why.
